### DISCUSSION TITLE
../submodule-name

### DISCUSSION THREAD

**Discussion by snan** [community/community#24001] [New to GitHub]
<p>Hi everyone! Can someone help me figure out a best practices to send pull reqs to a huge project with tons of submodules?</p>
<p>Specifically, <a href="https://github.com/mudita/MuditaOS/pull/1537#issuecomment-791897297" rel="noopener nofollow ugc">here is the issue</a>.</p>
<blockquote>
<p>In <code>.gitmodules</code>, the path to each module… let’s use lwip as an example. It’s <code>../lwip.git</code>.</p>
<p>So when I try to <code>git submodule update --init --recursive</code> in my own GitHub forked repo, it looks for a repo at <a href="https://github.com/snan/usb_stack.git%60" rel="noopener nofollow ugc">https://github.com/snan/usb_stack.git`</a> which doesn’t exist because GitHub didn’t recursively fork everything, just the main repo.</p>
</blockquote>
<p>Now, I’m trying to be careful to not “X Y problem” myself here. I wanna solve this problem the right way; just not sure what the right way is. The reason I need the submodules pulled is to be able to build and run tests. I haven’t started hacking on the submodules specifically yet (if/when I do, that’s gonna be a problem also).</p>

**Comment by airtower-luna** [comment]
<p>If I understand your description correctly, the upstream repository is configured with relative submodule URLs that assume that the submodule repository exists in the same user/organization account. I think that’s a questionable decision on the upstream project’s part, although for private projects it might be the easier approach because that makes the submodules use the same protocol and authentication mechanism as the main repo.</p>
<p>The <a href="https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-add-bltbranchgt-f--force--nameltnamegt--referenceltrepositorygt--depthltdepthgt--ltrepositorygtltpathgt"><code>git submodule add</code> documentation</a> describes how relative paths are resolved. In particular:</p>
<blockquote>
<p>The default remote is the remote of the remote-tracking branch of the current branch.</p>
</blockquote>
<p>So I suggest you try checking out a branch that tracks a branch in the upstream repo, and initializing the submodules from there.</p>

**Comment by snan** [comment]
<blockquote>
<p>If I understand your description correctly, the upstream repository is configured with relative submodule URLs that assume that the submodule repository exists in the same user/organization account.</p>
</blockquote>
<p>That’s right.</p>
<p>Everything works when I check out upstream.</p>
<p>But to be able to make pull requests I need to be able to work on a fork (right?).</p>
<p>Someone else gave me the suggestion to first download their repo (recursively with submodules) to my dev desktop box, and then add my fork of the main project as a remote and fetch it. But that didn’t work because I get:</p>
<pre><code class="lang-auto">Fetching submodule module-bsp/board/rt1051/bsp/usb
fatal: remote error: upload-pack: not our ref c29f13485f85fc3c10c0ac44ca1bd33fc10d2e7c
fatal: the remote end hung up unexpectedly
</code></pre>

**Comment by snan** [comment, marked as answer]
<p>Never mind, I just realized that I should just push to upstream. They are doing things old school, not using forks.</p>

**Comment by airtower-luna** [comment]
<p>I’m glad you could work around the problem!</p>
<p>I’m just realizing though: For repositories on GitHub using relative paths should work by going two levels up, so the relative paths include the owner of the submodule. Maybe you can suggest that as an improvement. :wink:</p>
